### ISSUE TITLE
Caret suddenly lost all content of saved & autosaved document left open for days

### ISSUE THREAD

**Issue by mjb2010** [thomaswilburn/Caret#353]
I left Caret running on my desktop for several days with a document open. The document had been explicitly saved to a local folder already, and its file name was showing on the tab, so I wasn't worried about losing it. autosaveInterval=1, so Caret dutifully auto-saved it every minute as well. Last night I went to sleep and the document was fine. Today I wake up, log in, and the file name is still in the tab, but the editor window is empty. In Explorer, the file shows as 0 bytes. I tried using Undo, but nothing came back—it was not an accidental edit; Caret simply lost the file contents. How could this have happened?

The local folder that the files were saved to is nothing special, just a folder on an internal SATA drive, same place I have all my documents.

Win7 64-bit, Chrome 41.0.2272.89 m (64-bit), Chrome last self-updated about 5 days ago, last defrag run was 5 days ago.

**Comment by thomaswilburn** [comment]
I don't know. Are you sure that Caret did it?

If you can provide more info, I'll look into it, but right now I can't think of an explanation that would help me debug.

**Comment by mjb2010** [comment]
I ruled out all the possibilities I could think of. It couldn't have been an accidental edit, because the undo buffer is empty. Chrome didn't update itself in the background. The disk isn't a USB drive and didn't go offline. No defrag ran, so there was no conflict with Caret's autosave. Nothing of note is in the system logs. No power loss or reboot.

The only other thing I do is once or twice a day, I switch between being seated at the console and logging in via Remote Desktop. At night, I end my Remote Desktop session, and in the morning I sit at the console and enter my password to unlock the screen. With each switch between modes (either starting a Remote Desktop session or starting a console session), Windows redraws the desktop. I believe this involves sending a signal to all the open applications to repaint their windows. But I don't think this is a problem for Chrome/Caret, because I do this every day and Caret didn't lose anything. Only today did I find the document window was empty.

Maybe Caret's autosave should do a sanity check, like if no changes have been made since the last autosave, verify that the document size about to be autosaved is the same as the size in the last autosave. If it's not, stop autosaving and prompt the user.

**Comment by thomaswilburn** [comment]
I don't know of any way to do that without falling into a heuristic mineshaft--any indication that the file size has changed in an undesirable way could also simply result from user action. For example, I maintain a to-do list on my desktop in most of my editors, and from time to time I wipe it out. Should Caret second-guess that? I don't particularly think it should--I think that'd be super annoying to have to double-confirm file saves.

Chrome presumably didn't update, but I did push out an update to Caret this weekend. It shouldn't restart on its own--Chrome should only replace the app when it's closed and re-opened, but that's the only thing I can think of.

Look, here's the thing. I'm sympathetic, because nobody likes losing their data. But I don't use the autosave myself, and I don't have any steps in this description on how to reproduce this bug. The autosave system uses the same mechanics as a regular save, so if it was actually a problem with the filesystem API we'd have seen it occur elsewhere. And when I leave files open for weeks at a time, I've never seen them obliterated. So I'm very interested in investigating what went wrong for you, but I'm going to need something to narrow it down. Adding additional nag messages to Caret (all of which would need to be tested, and might interact with the rest of the system poorly) in the hope that it will catch this kind of thing is not a viable solution at this time.

**Comment by mjb2010** [comment]
My suggestion was that Caret wouldn't do a size check if the user has done anything. The idea is to detect when the buffer empties *without* user interaction. Under what circumstances would that ever intentionally happen and be no cause for alarm?

Re: the update, I've still got it open on my desktop, and About says the version is 1.4.40. When I click Check For Updates, nothing happens.

**Comment by thomaswilburn** [comment]
How do I know when it's been emptied without user interaction? Lots of functionality in Caret sets the contents of the buffer, external messaging API notwithstanding. There's not a good way to know whether the user did something directly, or if they initiated another action that, for whatever reason, caused the file to empty and/or change radically. Setting up save dialogs that trigger on nebulous circumstances defeats the purpose of having an autosave.

Meanwhile, it's also possible for another process to update a file, in which case if there are no user modifications it will be reloaded. We don't want to mess with that either, because it's important: you might be monitoring a build process, or using a mounted drive, or pulling from source control and want the latest files to be brought into Caret.

Basically: if you can give me a reproducible test case, I'm happy to look at this. But until then, there's nothing I can do to solve it, and I'm not going to add UI to Caret based on assumptions that I don't think will hold true for many of my users.

**Comment by mjb2010** [comment]
Let's say I am able to reproduce this. Surely the steps would be exactly the same as I have already described. I open Caret, type something, and use Ctrl+S to save it to a file. I don't touch the window for 3 or 4 days, and then one day the content is gone. So, do you want to hear from me when that happens? I mean, I don't think you are going to be any more satisfied with what I can tell you about it than you are right now.

You seem to be annoyed that I reported anything at all, or that I should have known that it wouldn't be feasible to detect whether it's an internal Caret problem or the result of an external force acting on the file.

I'm not trying to waste your time or make you write code you don't want to write, and I wouldn't expect that any code you did write would necessarily be something you pushed out to all users; maybe you have a debug branch I don't know about. 

I am just bringing this issue to your attention. Maybe the fact that it has been reported will prompt others to contribute some "me too" responses, if they know where to look. Until then, sorry for wasting your time.
